design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+    `status.currentImage` and `status.currentHash` do not match the
+    `spec.package` and does not match an existing install `Job` (or existing
+    install `Job` does not exist), the controller will proceed with update.
+2. Delete any existing install `Job` with the same name as the `Package` and
+   create a new one with image in `spec.package`.
+  - If the `sha256` of the package image matches that in the `spec.hash` field
+    of the `Package` then no further action is needed, assuming the current
+    `PackageRevision` is healthy. This will only occur if the image tag changed
+    but the contents did not, and any previous install `Job` was manually
+    deleted.
+3. Check to see if a `PackageRevision` exists that has name equivalent to the
+   `sha256` hash of the image and is labeled with the `Package` name.
+   - If one does and it is `Active`, no further operation is needed.
+   - If one does and it is `Inactive`, mark it as `Pending` and continue.
+   - If one does and it is `Pending`, continue.
+   - If one does not, create a new `PackageRevision` (or locate an existing one)
+     with a name that matches the `sha256` of the image in `spec.package`, is
+     owned by the `Package`, has a `spec.state` of `Pending`, and is labeled
+     with the name of the `Package`.
+    - Note: If there are other `PackageRevision`s owned by this `Package` that
+      have state `Pending` they should be marked as `Inactive`. This could occur
+      if a previous upgrade was deemed unsafe or a `Package` was updated
+      multiple times in quick succession.
+4. Compare the `Pending` `PackageRevision` to the `Active` `PackageRevision` and
+   determine if it is safe to upgrade. If it is not safe to upgrade, do not
+   proceed. User will be required to change `spec.package` back to previous
+   version if they want to cancel upgrade. Until doing so, the previous
+   `PackageRevision` will continue to be `Active`.
+  - Safe upgrade will initially only be a check that support for CRDs for which
+    instances currently exist or are depended on by other `Package`s will not be
+    dropped, but can be expanded to be more sophisticated in the future.
+5. If upgrade is deemed safe, the `Active` `PackageRevision` should be marked as
+   `Stopping`, and controller should wait until the controller deployed by that
+   `PackageRevision` has been stopped (i.e. `Deployment` no longer exists).
+6. When the `Stopping` `PackageRevision` has finished cleaning up its
+   controller, it should be marked as `Inactive` and the new `PackageRevision`
+   should be marked as `Active`. This will cause it to install all of its
+   objects (updating any existing ones), and start its own controller (i.e.
+   create a `Deployment`).
+7. Wait for `PackageRevision` to become `Ready` (this will occur after
+   successful install of all objects included in the `PackageRevision`).
+8. Set the `status.currentHash` of the `Package` to match the `sha256` hash of
```

Comment:
remind me how we track the currently active package revision, is it just by querying all packageRevisions that are for a particular package and finding the 1 Active one?

Does the Package point directly to its currently active PackageRevision at all?

Reply:
Because the `PackageRevision` has a name equivalent to the hash of the image, we can find it by using the `status.currentHash` of the `Package`. If we somehow lost the status of the `Package` (e.g. in a backup and restore scenario), we would find the `PackageRevision` again by creating a `Job` seeing the the hash reported by it, then seeing that a `PackageRevision` already exists with that name 👍